<img style="display: block; margin-left: auto; margin-right: auto" src="./ccsf-logo.png" width="250rem;" alt="The CCSF black and white logo">

# Lab 01: Jupyter and Python

## Welcome 🎉

Welcome to your first Foundations of Data Science lab assignment! In almost every week, you will complete a lab assignment like this one. Most people cannot learn technical subjects without hands-on practice, so labs are an important part of the course! You will make mistakes as you complete the assignments in this course and that is okay! We are here to support you as you learn from those mistakes.

## Lab Assignment Reminders

- For all tasks indicated with a 🔎 that you must write explanations and sentences for, provide your answer in the designated space and check your response with a classmate, a tutor, or the instructor before moving on.
- Throughout this assignment and all future ones, please be sure to not re-assign variables throughout the notebook! _For example, if you use `max_temperature` in your answer to one question, do not reassign it later on. Otherwise, you will fail tests that you thought you were passing previously!_
- Collaborating on labs is more than okay -- it's encouraged! You should rarely remain stuck for more than a few minutes on questions in labs, so ask an instructor or classmate for help. (Explaining things is beneficial, too -- the best way to solidify your knowledge of a subject is to explain it.) Please don't just share answers, though.

*View the related <a href="https://ccsf.instructure.com" target="_blank">Canvas</a> Assignment page for additional details.*

## Jupyter Notebooks

<img style="display: block; margin-left: auto; margin-right: auto" src="jupyter-main-logo.svg" alt="Jupyter's main logo" width="100rem;">

This web application is called a Jupyter notebook and is part of [Project Jupyter](https://jupyter.org/). A notebook is a place to host share explanatory content intended for humans usually in association with computational programs and their results. This notebook format supports reproducible computational research where anyone with access to the notebook and the relevant data will more than likely be able to reproduce the work done by the original researchers. Reproducibility is part of the backbone of modern scientific research. Notebook files on a computer end in `.ipynb`. This file extension is a reference to the IPython programming environment developed by [Fernando Pérez](https://en.wikipedia.org/wiki/Fernando_P%C3%A9rez_(software_developer)). You'll notice that these notebooks can be edited and run on several different platforms such as the [CCSF JupyterHub](https://ccsf.cloudbank.2i2c.cloud/), [Google Colab](colab.research.google.com), [VS Code](https://code.visualstudio.com/), etc. The different platforms allow you to interact with the notebooks in slightly different ways, but the main components of a notebook typically persist through those differences.

<img style="display: block; margin-left: auto; margin-right: auto" src="otter-logo-smaller.png" alt="Otter-Grader's logo" width="100rem;">

We will utilize notebooks for lecture demonstrations and assignments such as this one. Typically, you will need to provide explanatory content or code for the associated **Tasks** throughout the notebook. To assist in grading your work, we will use an auto-grading tool called [Otter-Grader](https://otter-grader.readthedocs.io/en/latest). You'll learn about that grader later in this assignment. Let's get started by exploring the various types of cells you will work with in this course.

## Cells

From our point of view, notebooks consist of building blocks called cells. If you navigate through the document, you'll notice that the various cells are outlined by a rectangle. In terms of the basics of cells, you will work with Markdown cells and code cells.

### Markdown Cells

Markdown cells are typically where you will find the content intended for the explanatory material. There could be text, images, etc. in these cells to explain what is going on in the notebook and in the code cells. The text in these cells are formatted with a language called Markdown. You don't need to learn all the details of Markdown, but you might find it helpful to go through the tutorial on [markdowntutorial.com](https://www.markdowntutorial.com/) or reference the [Markdown Cheat Sheet](https://www.markdownguide.org/cheat-sheet/). 

* If you double-click a Markdown cell, then you'll be able to see how it was formatted. Sometimes we lock the Markdown cells in these notebooks, so you won't accidentally modify our explanatory content or instructions. You should have no program editing cells that you create or cells that are associated with the assignment tasks.
* To return back to the formatted version of the cell, select the ▶ button on the toolbar.

#### Task 01 📍🔎

<!-- BEGIN QUESTION -->

For your first task, edit the following Markdown cell and replace the text `_Type your answer here, replacing this text._`, with your name and a song, movie, book, etc. that you have been enjoying recently. (It is okay and preferred that you delete the placeholder text and replace it with your response.)

It is not required, but if you want to practice using Markdown, try to format your response as an unordered list with your name in **bold** and the name of the song, movie, book, etc. in _italics_. If you need to, review the [Markdown Cheat Sheet](https://www.markdownguide.org/cheat-sheet/).

Movies that are based on a video game version of Harry Potter. Hosted on Youtube.

<!-- END QUESTION -->

**Task 01 is marked with a 🔎 symbol, so it is manually checked and not auto-graded. For lab tasks like this, check your response with another student, a tutor, or the instructor before moving on.**

### Code Cells

Code cells host text associated with some programming language. We will utilize the Python 3 programming language in this course, and you will learn some of the basics of that language throughout the semester. The following cell is a Code cell. You'll notice that Code cells have a different structure than Markdown cells. You can run the Code cell by selecting it and pressing the ▶ button.

As you progress, you'll want to familiarize yourself with a few shortcuts rather than selecting the buttons. For example, to run a Code cell, you can hold down <kbd>Shift</kbd> and press <kbd>Return/Enter</kbd>.

In [ ]:
print("Hello, World!")

If you ran the above cell, you should notice that next to the Code cell is a set of square brackets with a number inside of it. 
* If a Code cell has not been run, then it will not contain a number `[ ]:`.
* Depending on how long it takes for the code to run, you'll see an asterisk within the brackets while the code is running `[*]:`.
* If it is the first cell to be executed in the notebook, then it will have a 1 in it `[1]:`. 
* The next cell (including the same cell) that is run will have a 2 within the brackets `[2]:`.
* The numbers in the code cells provide us with the sequence in which the Code cells were run. It is important to know which cells have been executed because Code cells ran later in time might depend on Code cells previously executed.

#### Output

Notice that once the above code cell was run, `Hello, World!` appeared below the Code cell as output. The output of a code cell contain images, data tables, etc. produced by the output from running the Python commands.

### Otter: An Auto-Grader

In every assignment, we will run some checks on your code to see if you are doing things correctly. That auto-grading is done with the help of the [Otter-Grader](https://otter-grader.readthedocs.io/en/latest/) library. Otter runs tests on your code to help check your progress. Sometimes these tests show you what we are testing for and what the results are, but there will be times when we run hidden tests after you submit your work to help grade for correctness. 

* The lab assignments in this course **do not** have hidden tests.
* The homework and project assignments **do** have hidden tests.

For every assignment after this one, you'll run a cell at the top of the notebook that initializes Otter.

#### Task 02 📍

In every notebook assignment, you will need to run a Code cell that starts with `# Initialize Otter`. 

The keyword that we will use as a default to use the auto-grader is `grader`.  By default, Python doesn't understand what the word `grader` means, so you need to run a code cell that imports the Otter grader into the notebook environment and assigns the name `grader` to the auto-grader and its configuration for this assignment.

For this task:

1. Run the following Code cell that starts with `# Initialize Otter`. (Before you run the cell,)
2. Run the grader.check("task_02") cell. 

In [5]:
# Initialize Otter
import otter
grader = otter.Notebook("lab01.ipynb")

In [6]:
grader.check("task_02")

task_02 results: All test cases passed!
task_02 - 1 message: ✅ You've successfully run the cell and initialized Otter!

## The Kernel

The kernel is a program that executes the code inside your notebook and outputs the results. We running a Python 3 kernel in this course. In the top right of your window, you can see a circle that indicates the status of your kernel. If the circle is empty (⚪), the kernel is idle and ready to execute code. If the circle is filled in (⚫), the kernel is busy running some code. If the circle has a lightning bolt through it, then the kernel has shut down for some reason and needs to be restarted.

As we described above, next to every code cell, you'll see some text that reads `[ ]:`. Before you run the cell, you'll see `[ ]:`. When the cell is running, you'll see `In [*]`. If you see an asterisk (\*) next to a cell that doesn't go away, it's likely that the code inside the cell is taking too long to run, and it might be a good time to interrupt the kernel (discussed below). When a cell is finished running, you'll see a number inside the brackets, like so: `In [1]`. The number corresponds to the order in which you run the cells; so, the first cell you run will show a 1 when it's finished running, the second will show a 2, and so on. 

### Restarting the Kernel

You may run into problems where your kernel is stuck for an excessive amount of time, your notebook is very slow and unresponsive, or your kernel loses its connection. If this happens, try the following steps:

* At the top of your screen, click `Kernel` in the menu bar, then `Interrupt Kernel`.
* If that doesn't help, click `Kernel` in the menu bar, then `Restart Kernel`. **If you do this, you will have to run your code cells from the start of your notebook up until the cell where you paused your work.**
* If that doesn't help, restart your server by completing the following tasks:
    1. Open [your JupyterHub home page](https://ccsf.cloudbank.2i2c.cloud/hub/home).
    2. Select the `Stop My Server` button, wait a few moments, and select the `Start My Server` button.
    3. Close the JupyterHub window/tabs.
    4. Re-open the notebook using the link from the Canvas assignment page. *You'll need to run your code cells again*.
    
**Note that if your notebook is inactive for a while, then the kernel will shut down on its own and you will have to run your code cells from the start of your notebook up until where you paused your work.**

### Task 03 📍🔎

<!-- BEGIN QUESTION -->

It is common for students to start an assignment, make some progress, take a break, and then return to their assignment. The kernel will likely shut down during the break to save resources. This might cause more confusion than you might think because the notebook will show your work is still there, and the code cells look like they have been run, but the Python interpreter doesn't remember anything once the kernel shuts down. So, you must re-run the cells you've already run before moving on. We want you to replicate this (common) situation in this task. 

For this task:

1. At the top of your screen, click `Kernel` in the menu bar, then click `Shut Down Kernel`. After you do this the Kernel status icon should show a lightning bolt and read No Kernel. You are doing this to simulate the kernel shutting down after you have made some progress on a notebook.
2. Next, refresh the browser page as if you were you returning to this assignment from a break and scroll back to here to continue working on this task.
3. After you refresh the page, the kernel should have restarted. (If it didn't, you can manually do it by clicking `Kernel` in the menu bar, then clicking `Restart Kernel...`.) If you go back up to the `grader.check` cell for Task 02 and try to run it, you should see that it fails even though you already ran the `# Initialize Otter` cell. This is because the keyword `grader` was removed from memory once the kernel shut down. Go try this and re-run the `# Initialize Otter` cell to pass Task 02 again.

<!-- END QUESTION -->

This task might feel a little complicated, so you might want to try it will a tutor or the instructor. With that said, do not skip it because you will eventually have to deal with this situation, so we recommend that you learn about it while you have support.

**Check your understanding of this task with another student, a tutor, or the instructor before moving on.**

## Adding and Editing Cells

### Adding Cells to Jupyter Notebooks

So far, you have only edited the content of this notebook. There will be times when you want to add your own cells to notebooks. To add a cell, click the + button in the toolbar.  It'll start out as a Code cell.  In order to change the cell type:

1. Click to the cell that you want to change the type of.
2. Click the drop-down box in the toolbar and choose `Markdown`.

### Making Mistakes

Before you begin the next task, we want to remind you that you will make mistakes as you work with these notebooks. Consider the following three pieces of advice:

1. If you find that you have just made a mistake, be careful of compounding the mistakes by trying to fix them. Instead, select `Edit` in the menu bar and select a relevant `Undo` option. You can undo changes made to notebook and to the individual cells.
2. You can create checkpoints in your notebook's history by selecting `File` in the menu bar, then `Save`. You can revert to the last checkpoint by selecting `File` in the menu bar, followed by `Revert Notebook to Checkpoint`.
3. If you feel that you just want to start your assignment over, then select `File` in the menu bar, followed by `Rename ...`. This keeps your work in your JupyterHub. After that, go back to Canvas and open the link to the notebook from the assignment page. This will create a new copy of the original notebook that we shared with you.

### Task 04 📍🔎

<!-- BEGIN QUESTION -->

Okay, it's time to practice adding cells and moving them around. For this task,

1. Add a code cell to the notebook and **position it below** this Markdown cell.
2. Type `2 + 2` in the code cell and run it so that `4` appears in its associated output.
3. Add a Markdown cell to the notebook and **position it above** the code cell you created.
4. Type `I'm getting the hang of this! Here is some code:` in the Markdown cell that you added.

I'm getting the hang of this! Here is some code:

In [4]:
2+2

4

<!-- END QUESTION -->

**You might need to ask for help when it comes to moving the cells around. Check with someone that you completed this task correctly and don't give up!**

### Edit and Command Modes

Notebooks have two modes. In edit mode, you can edit text in cells. Command mode is used for navigating the notebook using keyboard shortcuts.

- If the notebook is in command mode, press `enter` on the keyboard to change to edit mode.
- If the notebook is in edit mode, press `esc` on the keyboard to change to command mode.

It can be tricky at first understanding the difference between the two modes, but you will get the hang of it over time!

## Shortcuts

After a few lessons, you will mind might end up moving quicker than you can edit the notebook. Learning a few shortcuts can help you speed up your editing process. You can find most of the notebook book shortcuts by selecting `Help` in the menu bar and clicking `Notebook Shortcuts`.

A few of the most common shortcuts that we use are:

1. Hold `shift` while pressing `return` on the keyboard: This shortcut runs the selected cell and moves to the next cell.
2. While in command mode, pressing `b` on the keyboard will create a new cell below the selected cell.
3. While in command mode, pressing `d` on the keyboard twice will delete the selected cell.
4. While in command mode, pressing `m` on the keyboard turns the selected cell into a Markdown cell.

### Running Many Cells Automatically

There will be times when you need to re-run your cells. For example, you stopped working on an assignment and came back to it later, but the kernel had already shut down. If you want to quickly catch up to where last were, then 

1. Select the cell you last worked on.
2. Select `Kernel` from the menu bar, then `Restart Kernel and Run up to Selected Cell ...`.

Note that the cells will run in order from top to bottom and stop by executing the selected cell **unless** the kernel encounters an error. If there is an error in any of the code cells above the one you selected, then the process will stop at that point. You should probably address that error before moving one since most of our work depends on what has been done in previous cells.

## Errors

Python is a language, and like natural human languages, it has rules.  It differs from natural language in two important ways:

1. The rules are relatively _simple_.  You can learn most of them in a few weeks and gain reasonable proficiency with the language in a semester.
2. The rules are _rigid_.  If you're proficient in a natural language, you can understand a non-proficient speaker, glossing over small mistakes.  A computer running Python code is not smart enough to do that.

Whenever you write code, you'll make mistakes.  When you run a code cell that has errors, Python will sometimes produce error messages to tell you what you did wrong.

**Errors are okay; even experienced programmers make many errors.** When you make an error, you just have to find the source of the problem, fix it, and move on.

We have made an error in the next cell.  Run it and see what happens.

In [6]:
print("This line is missing something.")

This line is missing something.


Depending on the version of Python being used to run the notebook, the error might look slightly different. We are using version 3.10 of Python, so the error should show `SyntaxError: incomplete input` at the bottom of the output. Additionally, there should be a ^ symbol attempting to point to the issue. In this case, the issue is that there is a missing parenthesis.

There's a lot of terminology in programming languages, but you don't need to know it all in order to program effectively. If you see a cryptic message like this, you can often get by without deciphering it. (Of course, if you're frustrated, ask for help.)

### Task 05 📍🔎

<!-- BEGIN QUESTION -->

For this task, complete the following:

1. Edit the above cell containing the code `print("This line is missing something."` to fix the `SyntaxError`.
2. Re-run the cell and notice that the error message is replaced with the correct output.

<!-- END QUESTION -->

Check out [the official Python 3.10 documentation on Exceptions](https://docs.python.org/3.10/library/exceptions.html#concrete-exceptions) to view the types of errors you could see. Note that this documentation is a bit overwhelming to newcomers

## Expressions

The fundamental building blocks of Python code are expressions. In these notebooks, cells can contain multiple lines with multiple expressions. When you run a cell, the lines of code are executed in the order in which they appear. Every `print` statement prints a line. Run the next cell and notice the order of the output.

In [7]:
print("First this line is printed,")
print("and then this one.")

First this line is printed,
and then this one.


### Numbers

Quantitative information arises everywhere in Data Science. In addition to representing commands to print out lines, expressions can represent numbers and methods of combining numbers. The expression `3.2500` evaluates to the number `3.25`. Run the cell and see the result.

In [8]:
3.2500

3.25

Notice that we didn't have to `print`. When you run a notebook cell, if the last line has a value, then Jupyter helpfully prints out that value for you. However, it won't print out prior lines automatically.

In [9]:
print(2)
3
4

2


4

Above, you should see that 4 is the value of the last expression, 2 is printed, but 3 is lost forever because it was neither printed nor last.

You don't want to print everything all the time anyway.  But if you feel sorry for 3, change the cell above to print it.

### Arithmetic Expressions

The line in the next cell subtracts.  Its value is what you'd expect.  Run it to see the result.

In [10]:
3.25 - 1.5

1.75

Many basic arithmetic operations are built into Python.  The textbook section on [Expressions](http://www.inferentialthinking.com/chapters/03/1/Expressions.html) describes all the arithmetic operators used in the course.  The common operator that differs from typical math notation is `**`, which raises one number to the power of the other. So, `2 ** 3` stands for $2^3$ and evaluates to 8. 

#### Order of Operations

The [order of operations](https://en.wikipedia.org/wiki/Order_of_operations) is the same as what you learned in previous courses. Python also has parentheses to group operations.  For example, compare the outputs of the two cells below that contain the same numbers. The first cell uses parentheses to show 2023, but the second cell produces a different number because of the lack of parentheses!

In [ ]:
6 + (6 * 5 - (6 * 3)) ** 2 * ((2 ** 3) / 4 * 7) + 1

In [ ]:
6 + 6 * 5 - 6 * 3 ** 2 * 2 ** 3 / 4 * 7 + 1

In common Mathematics notation, the first expression is

$$6 + \left(6 \times 5 - (6 \times 3)\right)^2 \times \left(\frac{\left(2^3\right)}{4} \times 7\right) + 1,$$


while the second expression is

$$6 + 6 \times 5 - 6 \times 3^2 \times \frac{2^3}{4} \times 7 + 1.$$

#### Task 06 📍🔎

<!-- BEGIN QUESTION -->

Replace the ellipses (`...`) in the following code cell with a Python expression that evaluates to be the same value as:

$$5 \times \left(3 \frac{10}{11}\right) - 50 \frac{1}{3} + 2^{\left(0.5 \times 22\right)} - \frac{7}{33} + 7.$$  

That's five times three and ten elevenths, minus fifty and a third, plus two to the power of half twenty-two, minus seven thirty-thirds plus six.  As a reminder, $3 \frac{10}{11}$ means $3+\frac{10}{11}$, not $3 \times \frac{10}{11}$.

In [ ]:
...

<!-- END QUESTION -->

## Names and Assignment Statements

In natural language, we have terminology that lets us quickly reference very complicated concepts.  We don't say, "That's a large mammal with brown fur and sharp teeth!"  Instead, we might say, "That's a bear!"

In Python, we do a similar thing using _assignment statements_. An assignment statement has a name on the left side of an `=` sign and an expression to be evaluated on the right. Run the next cell that contains an assignment statement.

In [11]:
ten = 3 * 2 + 4

When you run that cell, Python first computes the value of the expression on the right-hand side, `3 * 2 + 4`, which is the number 10.  Then it assigns that value to the name `ten`.  At that point, the code in the cell is done running. It can be a little subtle to know if a code cell that contains an assignment statement has being executed. Don't forget that the number in `[...]: ` let's us know that the cell has been run. After running that cell, the value 10 is bound to the name `ten`. Take a look by running the next cell.

In [12]:
ten

10

The statement `ten = 3 * 2 + 4` is not asserting that `ten` is already equal to `3 * 2 + 4`. Rather, that line of code changes what `ten` means; it now refers to the value 10, whereas before it meant nothing at all. If the designers of Python had been ruthlessly pedantic, they might have made us write

    define the name ten to hereafter have the value of 3 * 2 + 4 

instead.  You will probably appreciate the brevity of "`=`"!  But keep that more long-winded expression in mind as you work with assignment statements.

### Task 07 📍

Replace the `...` in the following cell with an assignment statement where the name is `eleven` and the expression is any arithmetic expression that has a value of `11`.

In [11]:
eleven = 11

In [12]:
grader.check("task_07")

task_07 results: All test cases passed!
task_07 - 1 message: ✅ Great work Python recognizes eleven as a name!
task_07 - 2 message: ✅ The name eleven represents the numerical value 11!

It is common for people to assign a value to a name and then evaluate the name so that the value is displayed as output. This is demonstrated in the following cell.

In [ ]:
close_to_pi = 355 / 113
close_to_pi

Another common pattern is that a series of lines in a single cell will build up a complex computation in stages, naming the intermediate results. We encourage this type of thinking in this class! The next cell shows you an example.

In [ ]:
semimonthly_salary = 841.25
monthly_salary = 2 * semimonthly_salary
number_of_months_in_a_year = 12
yearly_salary = number_of_months_in_a_year * monthly_salary
yearly_salary

From the online learning platform W3Schools page [Python - Variable Names](https://www.w3schools.com/python/python_variables_names.asp):

> A variable can have a short name (like x and y) or a more descriptive name (age, carname, total_volume). Rules for Python variables:
> * A variable name must start with a letter or the underscore character
> * A variable name cannot start with a number
> * A variable name can only contain alpha-numeric characters and underscores (A-z, 0-9, and _ )
> * Variable names are case-sensitive (age, Age and AGE are three different variables)
> * A variable name cannot be any of the [Python keywords](https://www.w3schools.com/python/python_ref_keywords.asp).

Other than those rules, what you name something doesn't matter *to Python*.  For example, this cell does the same thing as the above cell, except everything has a different name. **However**, names are very important for making your code *readable* to yourself and others.  The cell above is shorter, but it's useless without an explanation of what it does.

In [10]:
a = 841.25
b = 2 * a
c = 12
d = c * b
d

20190.0

At some point, you'll start to care about the style of the code you are writing. It helps to follow a common style pattern when you are collaborating with others. PEP 8 is the style guide outlined in the Python documentation. That guide contains a lot of advanced vocabulary, but you should check out [the PEP 8 section Naming Conventions](https://peps.python.org/pep-0008/#naming-conventions) if you plan to work with others using the Python language.

### Task 08 📍

Consider the time interval between midnight January 1, 2012 and midnight January 1, 2022. We will guide you to calculating the number of seconds in this particular decade using descriptive variable names and intermediate steps.

For this task,
 1. Assign `number_of_days` to be the total number of days in the above time interval. Note that there are three leap years in this span of this particular decade (2012, 2016, 2020). A non-leap year has 365 days and a leap year has 366 days. 
 2. Using `number_of_days`, assign the name `seconds_per_decade` to the number of seconds between midnight January 1, 2012 and midnight January 1, 2022. You are welcome to define and use more intermediate variables such as `hours_per_day` in the calculation of `seconds_per_decade`.

In [7]:
number_of_days = (3 * 366) + (7 * 365)
hours_per_day = 24
seconds_per_decade = number_of_days * hours_per_day * 60 * 60
seconds_per_decade

315619200

In [8]:
grader.check("task_08")

task_08 results: All test cases passed!
task_08 - 1 message: ✅ You have assigned values to the names number_of_days and second_per_decade.
task_08 - 2 message: ✅ You have the correct number of days assigned to number_of_days!
task_08 - 3 message: ✅ You have the correct number of seconds in a decade!

## Comments

You may have noticed the first line in the code cell that initializes Otter earlier on in the notebook:

    # Initialize Otter
    
This is called a *comment*. It doesn't make anything happen in Python; Python ignores anything on a line after a `#`.  Instead, it's there to communicate something about the code to you, the human. Comments are extremely useful, but keep them to a minimum when using notebooks. Instead, utilize the Markdown cells to explain what your code is doing.

<img src="http://imgs.xkcd.com/comics/future_self.png" alt="A comic showing a conversation between a programmer and their future self through comments.">
Image Source: <a href="http://imgs.xkcd.com/comics/future_self.png">imgs.xkcd.com</a>

## Calling Functions

A function is a named block of code. The most common way to combine or manipulate values in Python is by calling functions. As a preview, functions essentially name a logical flow of code. Python comes with many built-in functions that perform common operations.

For example, Python's built-in absolute value function `abs` takes a single number as its argument and returns the absolute value of that number. Rather than working with the absolute value function code directly, you are more likely to reference that code by the name `abs`. Run the next two cells that contain a function call using the absolute value function `abs`.

In [18]:
abs(5)

5

In [19]:
abs(-5)

5

Note that `abs` is the name of the function, but `abs(5)` is the call expression because it sends an input of `5` to the function to be evaluated and the function returns a value. You will learn more about functions later in the course. For now, just practice using ones that are already defined.

### Multiple Arguments

Some functions take multiple arguments, separated by commas. For example, the built-in `max` function returns the maximum argument passed to it. Run the following cell to see how to make a function call using multiple arguments. 

In [20]:
max(2, -3, 4, -5)

4

## Understanding Nested Expressions

Function calls and arithmetic expressions can themselves contain expressions. You saw an example in the last question:

    abs(42-34)

has 2 number expressions in a subtraction expression in a function call expression.

Nested expressions can turn into complicated-looking code. However, the way in which complicated expressions break down is very regular.

Suppose we are interested in human heights that are very unusual.  We'll say that a height is unusual to the extent that it's far away on the number line from the average human height.  [An estimate](http://press.endocrine.org/doi/full/10.1210/jcem.86.9.7875?ck=nck&) of the average adult human height (averaging, we hope, over all humans on Earth today) is 1.688 meters.

So if an individual named Kayla is 1.21 meters tall, then their height is $|1.21 - 1.688|$, or $0.478$, meters away from the average.  Here's a picture of that:

<img src="numberline_0.png" alt="A number line showing the distance as the height.">

And here's how we'd write that in one line of Python code:

In [21]:
abs(1.21 - 1.688)

0.478

What's going on here? `abs` takes just one argument, so the stuff inside the parentheses is all part of that *single argument*.  Specifically, the argument is the value of the expression `1.21 - 1.688`.  The value of that expression is `-.478`.  That value is the argument to `abs`.  The absolute value of that is `.478`, so `.478` is the value of the full expression `abs(1.21 - 1.688)`.

Picture simplifying the expression in several steps:

1. `abs(1.21 - 1.688)`
2. `abs(-0.478)`
3. `0.478`

In fact, that's basically what Python does to compute the value of the expression!

### Task 09 📍

Say that Paola's height is 1.76 meters.  In the next cell, use `abs` to compute the absolute value of the difference between Paola's height and the average human height.  Give that value the name `paola_distance_from_average_m`.

<img src="numberline_1.png" alt="A number line visualizing the distance between Paola's height and an average human height.">

In [14]:
paola_distance_from_average_m = abs(1.688 - 1.76)
paola_distance_from_average_m

0.07200000000000006

In [15]:
grader.check("task_09")

task_09 results: All test cases passed!
task_09 - 1 message: ✅ You have assigned values to the name paola_distance_from_average_m.
task_09 - 2 message: ✅ You have the correct value for paola_distance_from_average_m

### More Nesting

Now, say that we want to compute the more unusual of the two heights.  We'll use the function `max`, which (again) takes two numbers as arguments and returns the larger of the two arguments.  Combining that with the `abs` function, we can compute the larger distance from average among the two heights. Just read through and run the following cell. Note that you are not expected to know how this more complex `print` statement works, but it is worth spending a few minutes trying to understand.

In [16]:
kayla_height_m = 1.21
paola_height_m = 1.76
average_adult_height_m = 1.688

# The larger distance from the average human height, among the two heights:
larger_distance_m = max(abs(kayla_height_m - average_adult_height_m), 
                        abs(paola_height_m - average_adult_height_m))

# Print out our results in a nice readable format:
print(f"The larger distance from the average height among \
these two people is {larger_distance_m} meters.")

The larger distance from the average height among these two people is 0.478 meters.


The line where `larger_distance_m` is computed looks complicated, but we can break it down into simpler components just like we did before.

The basic recipe is to repeatedly simplify small parts of the expression:
* **Basic expressions:** Start with expressions whose values we know, like names or numbers.
    - Examples: `paola_height_m` or `5`.
* **Find the next simplest group of expressions:** Look for basic expressions that are directly connected to each other. This can be by arithmetic or as arguments to a function call. 
    - Example: `kayla_height_m - average_adult_height_m`.
* **Evaluate that group:** Evaluate the arithmetic expression or function call. Use the value computed to replace the group of expressions.  
    - Example: `kayla_height_m - average_adult_height_m` becomes `-0.478`.
* **Repeat:** Continue this process, using the value of the previously-evaluated expression as a new basic expression. Stop when we've evaluated the entire expression.
    - Example: `abs(-0.478)` becomes `0.478`, and `max(0.478, 0.072)` becomes `0.478`.

You can run the next cell to load Google Slides where you can advance through a step-by-step presentation of that process. The slide navigation controls are at the bottom of the presentation window.

In [17]:
from IPython.display import IFrame
IFrame('https://docs.google.com/presentation/d/e/2PACX-1vTiIUOa9tP4p\
HPesrI8p2TCp8WCOJtTb3usOacQFPfkEfvQMmX-JYEW3OnBoTmQEJWAHdBP6Mvp053G/\
embed?start=false&loop=false&delayms=3000', 800, 600)

### Task 10 📍

Given the heights of players from the Golden State Warriors, write an expression that computes the smallest difference between any of the three heights. Your expression shouldn't have any numbers in it, only function calls and the names `thompson`, `curry`, and `wiggins`. Give the value of your expression the name `min_height_difference`.

We recommend that you compute the absolute difference between each pair of names that represent heights using `abs` and then find the smallest of those 3 absolute differences using `min`.

In [19]:
# The three players' heights, in inches
thompson =  78 # Klay Thompson is 6'6"
curry = 74 # Steph Curry is 6'2"
wiggins = 79 # Andrew Wiggins is 6'7"

thompson_curry = abs(thompson - curry)
thompson_wiggins = abs(thompson - wiggins)
curry_wiggins = abs(curry - wiggins)

min_height_difference = min (thompson_curry, thompson_wiggins, curry_wiggins)
min_height_difference
print("The smallest difference between any of the three heights is", 
      min_height_difference, 
      "inche(s).")

The smallest difference between any of the three heights is 1 inche(s).


In [20]:
grader.check("task_10")

task_10 results: All test cases passed!
task_10 - 1 message: ✅ You have assigned values to the name min_height_difference.
task_10 - 2 message: ✅ You have the correct value for min_height_difference

Great work getting to this point! That is enough for now to get you started working with Python. You will be learning a lot more over the semester. Now, complete the following steps to submit your assignment.

## Submit your Lab to Canvas

Once you have finished working on the lab questions, prepare to submit your work in Canvas by completing the following steps.

1. In the related Canvas Assignment page, check the requirements for a Complete score for this lab assignment.
2. Double-check that you have run the code cell near the end of the notebook that contains the command `grader.check_all()`. This command will run all of the run tests on all your responses to the auto-graded tasks marked with 📍.
3. Double-check your responses to the manually graded tasks marked with 📍🔎.
4. Select the menu items `File`, `Save and Export Notebook As...`, and `Html_embed` in the notebook's Toolbar to download an HTML version of this notebook file.
5. In the related Canvas Assignment page, click Start Assignment or New Attempt to upload the downloaded HTML file.

---

To double-check your work, the cell below will rerun all of the autograder tests.

In [21]:
grader.check_all()

task_02 results: All test cases passed!
task_02 - 1 message: ✅ You've successfully run the cell and initialized Otter!

task_07 results: All test cases passed!
task_07 - 1 message: ✅ Great work Python recognizes eleven as a name!
task_07 - 2 message: ✅ The name eleven represents the numerical value 11!

task_08 results: All test cases passed!
task_08 - 1 message: ✅ You have assigned values to the names number_of_days and second_per_decade.
task_08 - 2 message: ✅ You have the correct number of days assigned to number_of_days!
task_08 - 3 message: ✅ You have the correct number of seconds in a decade!

task_09 results: All test cases passed!
task_09 - 1 message: ✅ You have assigned values to the name paola_distance_from_average_m.
task_09 - 2 message: ✅ You have the correct value for paola_distance_from_average_m

task_10 results: All test cases passed!
task_10 - 1 message: ✅ You have assigned values to the name min_height_difference.
task_10 - 2 message: ✅ You have the correct value for 